src/train.py

```diff
@@ -0,0 +1,441 @@
+        if len(pred.shape) == 2:
+            assert pred.shape[1] == 1
+            pred.shape = (pred.shape[0], )
+        return score, loss, valid_index, pred
+
+    def is_finished(self):
+        return (len(self._losses) == self._num_cv
+                and len(self._scores) == self._num_cv)
+
+    def get_result(self):
+        if not self.is_finished():
+            print('not finished')
+            return
+        return np.mean(self._scores), np.mean(self._losses), self._out_of_fold_prediction
+
+
+class GPUController:
+    def __init__(self):
+        self._avail_gpu_list = [0, 3, 5, 6, 7]
+        self._queue = mp.Queue()
+        for i in self._avail_gpu_list:
+            self._queue.put(i)
+
+    def get_gpu_id(self):
+        gpu_id = self._queue.get()
+        return gpu_id
+
+    def release_gpu_id(self, gpu_id):
+        self._queue.put(gpu_id)
+
+    def max_size(self):
+        return len(self._avail_gpu_list)
+
+
+pd.set_option('display.max_columns', 5000)
+
+NUM_CV = 5
+N_BAGS = 8
+
+gpu_controller = GPUController()
+
+# Load data
+config = DatasetLoaderConfig()
+train_csv_dsl = DatasetLoader(config=config)
+train_csv_dsl.load(conf.TRAIN_DATASET)
+X_train_csv = train_csv_dsl.get_feature().values
+y_train_csv = train_csv_dsl.get_label().values
+
+config.toggle_read_mode(True)
+test_csv_dsl = DatasetLoader(config=config)
+test_csv_dsl.load(conf.TEST_DATASET)
+X_test = test_csv_dsl.get_feature().values
+
+out_of_fold_predictions_dict = {
+    'xgb': None,
+    'lgb': None,
+    'dnn': None,
+}
+
+
+# XGBoost K-fold validation
+print('XGBoost K-fold validation')
+
+
+def fit_xgb(i, X_train_cv, y_train_cv):
+    gpu_id = gpu_controller.get_gpu_id()
+    model = xgb.XGBRegressor(gpu_id=gpu_id, **xgboost_param)
+    model.fit(X_train_cv, y_train_cv)
+    gpu_controller.release_gpu_id(gpu_id)
+    return model
+
+
+runner = CVRunner(train_csv_dsl, NUM_CV, fit_xgb)
+runner.run(n_jobs=gpu_controller.max_size())
+
+mean_score, mean_loss, out_of_fold_prediction = runner.get_result()
+out_of_fold_predictions_dict['xgb'] = out_of_fold_prediction
+print('[xgb] score: {}, loss: {}'.format(mean_score, mean_loss))
+print(out_of_fold_predictions_dict['xgb'])
+
+
+# lightgbm K-fold validation
+print('lightGBM K-fold validation')
+
+
+def fit_lgb(i, X_train_cv, y_train_cv):
+    model = lgb.LGBMRegressor(**lgb_param)
+    model.fit(X_train_cv, y_train_cv)
+    return model
+
+
+runner = CVRunner(train_csv_dsl, NUM_CV, fit_lgb)
+runner.run(n_jobs=5)
+
+mean_score, mean_loss, out_of_fold_prediction = runner.get_result()
+out_of_fold_predictions_dict['lgb'] = out_of_fold_prediction
+print('[lgb] score: {}, loss: {}'.format(mean_score, mean_loss))
+print(out_of_fold_predictions_dict['lgb'])
+
+
+# DNN K-fold validation
+print('DNN k-fold validation')
+cols = train_csv_dsl.get_feature().columns
+
+
+def dnn_lr_schd(epoch, lr):
+    if epoch == 100 or epoch == 150:
+        return lr * 0.1
+    else:
+        return lr
```

Comment:
(optional) It's better to move this along with reduce_lr and model.fit to model.py and wrap these into a class.